rs-matter/src/data_model/core.rs

```diff
@@ -15,142 +15,802 @@
  *    limitations under the License.
  */
 
-use portable_atomic::{AtomicU32, Ordering};
+use core::cell::{Cell, RefCell};
+use core::iter::Peekable;
+use core::pin::pin;
+use core::time::Duration;
 
-use super::objects::*;
-use crate::{
-    alloc,
-    error::*,
-    interaction_model::core::Interaction,
-    transport::{exchange::Exchange, packet::Packet},
+use embassy_futures::select::select;
+use embassy_time::{Instant, Timer};
+use log::{debug, error, info, warn};
+
+use crate::acl::Accessor;
+use crate::interaction_model::messages::ib::AttrStatus;
+use crate::utils::buf::BufferAccess;
+use crate::{error::*, Matter};
+
+use crate::interaction_model::core::{
+    IMStatusCode, OpCode, ReportDataReq, PROTO_ID_INTERACTION_MODEL,
 };
+use crate::interaction_model::messages::msg::{
+    InvReq, InvRespTag, ReadReq, ReportDataTag, StatusResp, SubscribeReq, SubscribeResp, TimedReq,
+    WriteReq, WriteRespTag,
+};
+use crate::respond::ExchangeHandler;
+use crate::tlv::{get_root_node_struct, FromTLV, TLVWriter, TagType};
+use crate::transport::exchange::{Exchange, MAX_EXCHANGE_RX_BUF_SIZE, MAX_EXCHANGE_TX_BUF_SIZE};
+use crate::utils::writebuf::WriteBuf;
 
-// TODO: For now...
-static SUBS_ID: AtomicU32 = AtomicU32::new(1);
+use super::objects::*;
+use super::subscriptions::Subscriptions;
 
 /// The Maximum number of expanded writer request per transaction
 ///
 /// The write requests are first wildcard-expanded, and these many number of
 /// write requests per-transaction will be supported.
-pub const MAX_WRITE_ATTRS_IN_ONE_TRANS: usize = 7;
+const MAX_WRITE_ATTRS_IN_ONE_TRANS: usize = 7;
 
-pub struct DataModel<T>(T);
+pub type IMBuffer = heapless::Vec<u8, MAX_EXCHANGE_RX_BUF_SIZE>;
+
+struct SubscriptionBuffer<B> {
+    node_id: u64,
+    id: u32,
+    buffer: B,
+}
+
+/// An `ExchangeHandler` implementation capable of handling responder exchanges for the Interaction Model protocol.
+/// The implementation needs a `DataModelHandler` instance to interact with the underlying clusters of the data model.
+pub struct DataModel<'a, const N: usize, B, T>
+where
+    B: BufferAccess<IMBuffer>,
+{
+    handler: T,
+    subscriptions: &'a Subscriptions<N>,
+    subscriptions_buffers: RefCell<heapless::Vec<SubscriptionBuffer<B::Buffer<'a>>, N>>,
+    buffers: &'a B,
+}
 
-impl<T> DataModel<T> {
-    pub fn new(handler: T) -> Self {
-        Self(handler)
+impl<'a, const N: usize, B, T> DataModel<'a, N, B, T>
+where
+    B: BufferAccess<IMBuffer>,
+    T: DataModelHandler,
+{
+    /// Create the handler.
```

Comment:
nit: as it stands this just states the obvious (a method `new` returning `Self` is implicitly creating the object.

I would suggest to either drop the obvious doc or add things that are not obvious from the signature.

Reply:
Let's keep it. At some point - I hope - we'll enable `!#[forbid(missing_docs)]` in `lib.rs` and then you need _every_ public method or type to be documented. Even if with a single sentence. The alternative (`#[allow(missing_docs)]`) sprincled here and there somehow does not resonate with me.

I'll see if I can derive something more useful from the method parameters.

Reply:
Added extra comments describing the input parameters.